### PR TITLE
Docs: add DSO, GOT, and PLT to `Lexicon.md`

### PR DESCRIPTION
**Pull request by MaxDesiatov**
These terms are frequently referenced in Swift source code and comments. Having at least basic definitions for them in `Lexicon.md` should lower the barrier for new contributors.

### CODE REVIEW

File: docs/Lexicon.md [hunk 1]
```diff
@@ -188,6 +188,12 @@ guarantee this. This was
 "Do not merge". Placed in PR titles where discussion or analysis is still
 ongoing.
 
+## DSO
+
+Dynamic shared object, an object file to be used by multiple applications
+while they are executing. Dynamically linked libraries are implemented with
+DSOs.
```
Comment (compnerd) [inline docs/Lexicon.md:195]: The DSO is the .so/.dylib/.dll itself, not the object files that are used.

File: docs/Lexicon.md [hunk 2]
```diff
@@ -250,6 +256,15 @@ A representation of all generic parameters and their requirements. Like
 types, generic signatures can be [canonicalized](#canonical-type) to be
 compared directly.
 
+## GOT
+
+[Global offset table](https://en.wikipedia.org/wiki/Global_Offset_Table),
+a section in an executable or a shared library, which enables code to run
+independently of the memory address where it is loaded at runtime. Entries
+in GOT directly point to absolute addresses of symbols. The format of GOT
+is platform-dependent. Dynamic linkers update GOT relocations either on
```
Comment (compnerd) [inline docs/Lexicon.md:265]: I prefer loader to dynamic linker.

File: docs/Lexicon.md [hunk 3]
```diff
@@ -455,6 +470,12 @@ only to accelerate the process of reading C/C++/Objective-C headers, such as
 the bridging headers read in by the `-import-objc-header` command-line
 flag to swiftc.
 
+## PLT
+
+Procedure linkage table, which is used to call external functions that don't
+have their addresses known at link time. These addresses are then resolved
+by a dynamic linker at run time.
```
Comment (compnerd) [inline docs/Lexicon.md:477]: Similar

**Review by compnerd — COMMENTED** [review]
One thing that concerns me about this is that it can grow uncontrolled.  Each platform has its own lexicon for this.  Toolchain can use different lexicon as well.  These terms aren't Swift specific, but are general toolchain terms.  Do we intend to document all principles of toolchain work?

**Comment by MaxDesiatov** [comment]
> One thing that concerns me about this is that it can grow uncontrolled. Each platform has its own lexicon for this. Toolchain can use different lexicon as well. These terms aren't Swift specific, but are general toolchain terms. Do we intend to document all principles of toolchain work?

I personally don't intend to document all of them, but only those that I personally found hard to piece together, especially abbreviations. These three abbreviations are used extensively in the codebase and comments. It took me some time to understand their meaning and context in which they're used.

I don't know the original intention of the author of this file, but it looks like a precedent has been established with a lot of terms specified in this document not specific to Swift itself.

**Comment by CodaFi** [comment]
Perhaps we ought to refine/scope these entries to their usages in the project.  For example, IRGen makes heavy usage of "GOT" as in "GOT-equivalent variable" to describe the way it forms references to global objects that may or may not be external references. We can also cross reference resources like the LLVM langref if need be (for example, they provide a pragmatic explanation of DSO local variables).

**Comment by MaxDesiatov** [comment]
@swift-ci smoke test

**Comment by MaxDesiatov** [comment]
@swift-ci test windows

**Review by xedin — APPROVED** [review]

**Pull request merged**